### PR TITLE
Feat: Multi Node example

### PR DESCRIPTION
**Pull request by OlivierDehaene**
An example of a multi node PyTorch STL10 run:

- Distribute on multiple GPUs
- Distribute on multiple nodes

Output for the first epoch: 

```
1 - Number of nodes: 2
1 - Node ID        : 1
1 - Local rank     : 0
1 - Global rank    : 1
1 - World size     : 2
1 - GPUs per node  : 1
1 - Master         : False
1 - Multi-node     : True
1 - Multi-GPU      : True
1 - Hostname       : r8i6n0
Initializing PyTorch distributed ...
Initialize Model...
Initialize Dataloaders...
0 - Number of nodes: 2
0 - Node ID        : 0
0 - Local rank     : 0
0 - Global rank    : 0
0 - World size     : 2
0 - GPUs per node  : 1
0 - Master         : True
0 - Multi-node     : True
0 - Multi-GPU      : True
0 - Hostname       : r8i5n8
Initializing PyTorch distributed ...
Initialize Model...
Initialize Dataloaders...
Epoch: [0][0/40]        Time 0.199 (0.199)      Speed 30.196 (30.196)   Loss 2.5202379227 (2.5202)      Prec@1 10.938 (10.938)  Prec@5 42.188 (42.188)
Epoch: [0][10/40]       Time 0.027 (0.113)      Speed 220.952 (53.131)  Loss 3.5998477936 (3.0600)      Prec@1 12.500 (11.719)  Prec@5 57.812 (50.000)
Epoch: [0][20/40]       Time 0.026 (0.084)      Speed 227.559 (71.366)  Loss 3.1009304523 (3.0737)      Prec@1 13.281 (12.240)  Prec@5 41.406 (47.135)
Epoch: [0][30/40]       Time 0.026 (0.070)      Speed 227.988 (86.164)  Loss 2.9206233025 (3.0354)      Prec@1 9.375 (11.523)   Prec@5 62.500 (50.977)
Test: [0/63]    Time 0.004 (0.004)      Speed 1488.817 (1488.817)       Loss 2.2089750767 (2.2090)      Prec@1 17.188 (17.188)  Prec@5 66.406 (66.406)
 * Prec@1 15.625 Prec@5 65.737
Epoch Summary:
        Epoch Accuracy: 15.625
        Best Accuracy: 15.625
```

### CODE REVIEW

**Comment by OlivierDehaene** [comment]
@mdiazmel @ylabbe the code is ready for review. Working on a comprehensive readme.

File: examples/pytorch/distributed_stl10/stl10_example.py [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+
+    print("Initializing PyTorch distributed ...")
+    torch.distributed.init_process_group(
+        init_method='env://',
+        backend='nccl',
+    )
+
+    print("Initialize Model...")
+    # Construct Model
+    model = models.resnet18(pretrained=False, num_classes=10).cuda()
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/stl10_example.py:276]: The weights are randomly initialized here but it doesn't seem that they are synchronized across processes ? 
It can be done by either setting torch.seed in all processes to the same seed (ensure weights are initialized similarly in all processes) or by synchronzining them explicitly (e.g. save checkpoint to disk in rank 0 and load it in others).

File: examples/pytorch/distributed_stl10/stl10_example.py [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+    print(PREFIX + "Global rank    : %i" % global_rank)
+    print(PREFIX + "World size     : %i" % world_size)
+    print(PREFIX + "GPUs per node  : %i" % n_gpu_per_node)
+    print(PREFIX + "Master         : %s" % str(is_master))
+    print(PREFIX + "Multi-node     : %s" % str(multi_node))
+    print(PREFIX + "Multi-GPU      : %s" % str(multi_gpu))
+    print(PREFIX + "Hostname       : %s" % socket.gethostname())
+
+    # set GPU device
+    torch.cuda.set_device(local_rank)
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/stl10_example.py:266]: This can be problematic when training with a number of gpus != number of gpus of the nodes. An option is for example
`torch.cuda.set_device(int(os.environ['CUDA_VISIBLE_DEVICES'].split(',')[local_rank]))`

File: examples/pytorch/distributed_stl10/stl10_example.py [hunk 3]
```diff
@@ -0,0 +1,360 @@
...
+        correct = pred.eq(target.view(1, -1).expand_as(pred))
+
+        res = []
+        for k in topk:
+            correct_k = correct[:k].view(-1).float().sum(0, keepdim=True)
+            res.append(correct_k.mul_(100.0 / batch_size))
+        return res
+
+
+def reduce_tensor(tensor: torch.Tensor, world_size: int):
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/stl10_example.py:62]: Also possible to use this https://github.com/pytorch/vision/blob/1aef87d01eec2c0989458387fa04baebcc86ea7b/references/detection/utils.py#L118 when there are many metrics to reduce

File: examples/pytorch/distributed_stl10/stl10_example.py [hunk 4]
```diff
@@ -0,0 +1,360 @@
...
+
+    # define master address and master port
+    hostnames = subprocess.check_output(['scontrol', 'show', 'hostnames', os.environ['SLURM_JOB_NODELIST']])
+    master_addr = hostnames.split()[0].decode('utf-8')
+
+    # set environment variables for 'env://'
+    os.environ['MASTER_ADDR'] = master_addr
+    os.environ['MASTER_PORT'] = str(29500)
+    os.environ['WORLD_SIZE'] = str(world_size)
+    os.environ['RANK'] = str(global_rank)
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/stl10_example.py:245]: This can be simplified a little using `initfile` initialization method, but I am not sure what is the best practice.

File: examples/pytorch/distributed_stl10/stl10_example.py [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+
+    # Create DistributedSampler to handle distributing the dataset across nodes
+    # This can only be called after torch.distributed.init_process_group is called
+    train_sampler = DistributedSampler(trainset)
+    val_sampler = DistributedSampler(valset)
+
+    # Create the Dataloaders to feed data to the training and validation steps
+    train_loader = DataLoader(trainset,
+                              batch_size=batch_size,
+                              num_workers=10,
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/stl10_example.py:303]: I haven't run this specific example but in my code, I have observed some performance drop when using a number of workers equal to the number of cpus assigned to the task. I think having too many dataloading processes can hurt inter-node communication.

Regarding perforamnce, I have also observed that adding more interfaces can help, e.g. `export NCCL_SOCKET_IFNAME=ib0,bond1,bond0,eno1,eno2` but I am not sure if this would help for this simple MNIST example.

File: examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh [hunk 6]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+#SBATCH --job-name=pytorch_stl10     # job name
+#SBATCH --ntasks=2                   # number of MP tasks
```
Comment (ylabbe) [inline examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh:3]: What do you think about also adding an example that use one or two entire nodes ?

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 1 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:276]: It should be automatically handled inside the DistributedDataParallel initialization. See https://github.com/pytorch/pytorch/blob/master/torch/nn/parallel/distributed.py#L398.

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 3 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:62]: Thanks! I wasn't aware of this.

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 5 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:303]: >I haven't run this specific example but in my code, I have observed some performance drop when using a number of workers equal to the number of cpus assigned to the task. I think having too many dataloading processes can hurt inter-node communication.

Indeed, `num_workers` adds some overhead to data loading that might not be worth it if the transform pipeline is simple and the dataset is small (which is the case here) but I think its a good baseline to start with.

> Regarding perforamnce, I have also observed that adding more interfaces can help, e.g. export NCCL_SOCKET_IFNAME=ib0,bond1,bond0,eno1,eno2 but I am not sure if this would help for this simple MNIST example.

I didn't know about this. Maybe we should add it to the config?

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 2 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:266]: I don't think this is the case. The Slurm Generic Resource (GRES) Scheduling should handle this by setting the `CUDA_VISIBLE_DEVICES` for each local task according to the definition inside the `pytorch_distributed_stl10_example.sh` script:

```bash
#SBATCH --ntasks=2                         
#SBATCH --ntasks-per-node=1         
#SBATCH --gres=gpu:1                      # sets CUDA_VISIBLE_DEVICES
```

Comment (ylabbe) [reply examples/pytorch/distributed_stl10/stl10_example.py:266]: Oh, if CUDA_VISIBLE_DEVICES=3,4, torch.cuda.set_device(0) sets torch to use CUDA device 3 ? I always set CUDA_VISIBLE_DEVICES to be unique for each process but it is good to know !

**Comment by lesteve** [comment]
@ylabbe you are way more competent than me on this one, I invited you to the jean-zay-users doc and gave you admin rights so feel free to merge this one whenever you think it is ready.

File: examples/pytorch/distributed_stl10/README.md [hunk 7]
```diff
@@ -0,0 +1,6 @@
+# PyTorch STL10 example script
+
+You can launch the job (two single GPU nodes) with:
+```
+sbatch jean-zay-doc/examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh
```
Comment (mdiazmel) [inline examples/pytorch/distributed_stl10/README.md:5]: I propose this change:
```suggestion
cd jean-zay-doc/examples/pytorch/distributed_stl10/
sbatch ./pytorch_distributed_stl10_example.sh
```
And also the hard-coded folder in the `pytorch_distributed_stl10_example.sh` script (see next comment). This will avoid silly mistakes if someone clone the repository in a different place from  `$WORK`.

File: examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+#SBATCH --ntasks-per-node=1          # number of MPI tasks per node
+#SBATCH --gres=gpu:1                 # number of GPUs per node
+#SBATCH --cpus-per-task=10           # number of cores per tasks
+#SBATCH --hint=nomultithread         # we get physical cores not logical
+#SBATCH --time=00:10:00              # maximum execution time (HH:MM:SS)
+#SBATCH --output=pytorch_stl10%j.out # output file name
+#SBATCH --error=pytorch_stl10%j.err  # error file name
+
+set -x
+cd $WORK/jean-zay-doc/examples/pytorch/distributed_stl10
```
Comment (mdiazmel) [inline examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh:13]:
```suggestion
cd ${SLURM_SUBMIT_DIR}
```

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 5 above)
Comment (ylabbe) [reply examples/pytorch/distributed_stl10/stl10_example.py:303]: Hmm, I have recently performed some tests on a different task and I couldn't see a difference between using these variables or letting NCCL find the correct network interfaces by itself. Maybe the performance difference I noticed at the time came from something else. Therefore, I don't think it it necessary to add it in this PR, especially that it can cause errors if the specified interfaces are not available on some nodes (not all are available on all queues).

File: examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh (hunk 8 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh:13]: I agree with these changes. The only reason I did it this way was to stay consistent with the other torch example. We should push this changes to both example then.

**Comment by OlivierDehaene** [comment]
As of torch==1.6.0, mixed precision was integrated in the `torch.cuda.amp` submodule and doesn't require NVIDIA Apex anymore. Do you want me to patch this example to implement it?

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 2 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:266]: Yes! This is very useful in distributed.

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 4 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:245]: Environment variable initialization is the default method in torch. The initfile method can lead to unexpected behaviour if the previous run was unable to correctly clean the file and I find it a bit more complicated. If you want I can implement it.

File: examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh (hunk 8 above)
Comment (mdiazmel) [reply examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh:13]: Yes, you're right. I'll modify the other example to keep consistency.

**Comment by mdiazmel** [comment]
> As of torch==1.6.0, mixed precision was integrated in the `torch.cuda.amp` submodule and doesn't require NVIDIA Apex anymore. Do you want me to patch this example to implement it?

Not sure that Pytorch==1.6.0 is already available on Jean Zay (using `module`). Currently I only see:

```
pytorch-gpu/py3/1.2.0
pytorch-gpu/py3/1.3
pytorch-gpu/py3/1.3.0+nccl-2.5.6
pytorch-gpu/py3/1.3.1
pytorch-gpu/py3/1.3.1+nccl-2.5.6
pytorch-gpu/py3/1.4.0
pytorch-gpu/py3/1.5.0
pytorch-gpu/py3/1.5.1
```
Of course, It can be installed with `pip` or `Conda` but I think that, to make easy to reproduce these examples, it's better to keep using the tools already installed by JZ's admins.

I suggest to put a word on the readme (warning about this change in new Pytorch versions), merge in the current state and propose a new PR with the patch once torch==1.6.0 sill be available via `module`. What do you thing?

**Comment by OlivierDehaene** [comment]
Ok! Maybe we can leave the example as is and add the mixed precision to the "tips and tricks" section instead as it is very to use (see [tutorials](https://pytorch.org/docs/stable/notes/amp_examples.html#amp-examples)).

File: examples/pytorch/distributed_stl10/stl10_example.py (hunk 4 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/stl10_example.py:245]: Also, I'm not not sure how this would work as if we use the Initfile way, we need to know the master address ahead of time.

File: examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh (hunk 6 above)
Comment (OlivierDehaene) [reply examples/pytorch/distributed_stl10/pytorch_distributed_stl10_example.sh:3]: The README should now better reflect how to do that.

**Comment by OlivierDehaene** [comment]
I modified the README to add credit to the original torch example and better describe what the Slurm options do in this context.
